### ISSUE TITLE
why I got a no response when using codes like fitz.open(<filename>, <bytearry_stream>) to render a pdf file

### ISSUE THREAD

**Issue by Applex007** [rk700/PyMuPDF#173]
Hi JorjMcKie!
 
Reading the pdf file like 'stream = bytearray(open(<pdffile.pdf>, "rb").read())' and then using fitz.open(<filename>, stream) got a no response

Whether I used python version 2.7.14 or 3.6.5, this situation is same.

I also see the discussion of david-hughes and you on 27 Jul 2016, actually I first used wx.lib.pdfviewer of wxpython 4.0.1 to render and show my pdf files, I wrote the codes " pdfV.viewer.LoadFile(open(<pdffile.pdf>, "rb")" but it hints "Out of memory, Zoom level too heigh?",

By the way, my pdf file's size is 5263KB, I tried a file 497KB and it can show first page but other pages  usually couldn't rendering if frequenttly scrolling or zooming.

Any suggestions to help, thank you very much

**Comment by JorjMcKie** [comment]
Hi @Applex007 
Any chance to send me the file?
Just making sure I am getting you right: you shouldn't have a problem using this sequence of statements:

```python
stream = open(fname, "rb").read() # unimportant whether bytes or bytearray
doc = fitz.open("pdf", stream) # the string "pdf" is important to indicate document format
# now ready to work with doc ...
```
Please disclose your platform, PyMuPDF version and maybe also wxPython version.
Re wxPython, I also had issues with pdfviewer in earlier versions. Since I am using the Phoenix versions, that seems to be gone.

In earlier PyMuPDF versions, there also was a problem with not freeing memory of Pixmaps (which are the basis for display in wxPython). Every zooming and scrolling therefore increased the used memory ... until up to a certain point.

So, in summary, the more info you can provide me, the better.

**Comment by Applex007** [comment]
Hi JorjMcKie:

Thanks for your nice reply

As my question said, I used newest pymupdf 1.13.4 and my wxpython version is 4.0.1, the application I used is pycharm on win10

I just tried to modify the pdfviewer(wx) (see pdfdisplay(wx).py) and pdfdisplay.py of demo and tested them based a pure python3.6.5 environment, and the codes of modified and pdf test file will show you as an attachment

This time, when I running the pdfdisplay(wx).py, it can,t render and show errors as below when I scroll the pages, and running the pdfdisplay it still show no response and crash if I tried to scrolling

  `error: name too long 
  warning: cannot load content stream part 1/8 
  error: name too long 
  warning: cannot load content stream part 2/8 
  error: name too long 
  warning: cannot load content stream part 3/8 
  error: name too long 
  warning: cannot load content stream part 4/8 
  error: name too long 
  warning: cannot load content stream part 5/8 
  error: name too long
  warning: cannot load content stream part 6/8
  error: name too long
  warning: cannot load content stream part 7/8
  error: name too long
  warning: cannot load content stream part 8/8
  error: name too long
  warning: cannot load content stream part 1/8
  error: name too long
  warning: cannot load content stream part 2/8
  error: name too long
  warning: cannot load content stream part 3/8
  error: name too long
  warning: cannot load content stream part 4/8
  error: name too long
  warning: cannot load content stream part 5/8
  error: name too long
  warning: cannot load content stream part 6/8
  error: name too long
  warning: cannot load content stream part 7/8
  error: name too long
  warning: cannot load content stream part 8/8
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  warning: non-page object in page tree ()
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache
  error: name too long
  warning: cannot load object (1 0 R) into cache`

I’m not sure if I have a right way to use the pymupdf, the target What I want to got is open and show a pdf file from the memory on my GUI program (the file comes from an decrypt file stream)

Looking forward to your reply and thank for help very much again

[PDFTest.zip](https://github.com/rk700/PyMuPDF/files/2041887/PDFTest.zip)

**Comment by JorjMcKie** [comment]
I believe (without having looked too hard), that the PDF-in-memory object does not remain available.
And this is a **MUST-BE**, because MuPDF accesses the PDF several times again. So for example you cannot create `bio` in a function, open it as a PDF only return the `fitz.Document`.
Object `bio` is being dropped somewhere, I just haven't seen exactly where. The errors you sent look exactly like this. I bet that everything works fine if you use a conventional file instead.

**Comment by JorjMcKie** [comment]
I haven't the details of wx pdfviewer in my head, but if you can pass in `bio` into its `__init__()` and do the `LoadFile(bio)` there, it may work as you intend it.

**Comment by Applex007** [comment]
I tried to modified the `class mupdfProcessor` of `wxpython.lib.pdfviewer.viewer` as below and it worked fine

As well as, the modification of `class PDFdisplay` in pdfdisplay seems fine

In my opinion, the `bio` pass in `loadFile` or `fitz.open` shouldn't just a pointer variable but a really memeory copy variable, so it is not be interrupted when parse or render the file

Just my thought, I'm not a professional python programmer, anyway thanks for your help, this question 
seems resolved temporarily 

`wx.lib.pdfviewer.viewer.py`
class mupdfProcessor(object):
    def __init__(self, parent, pdf_file):
	# modified by applex
		...
		else:
			# assume it is a file-like object, pass the stream content to fitz.open
			# and a '.pdf' extension in pathname to identify the stream type
			pathname = 'fileobject.pdf'
			if pdf_file.tell() > 0:     # not positioned at start
				pdf_file.seek(0)
			**self.stream** = bytearray(pdf_file.read())
			self.pdfdoc = fitz.open(pathname, **self.stream**)


`pdfdisplay.py`
class PDFdisplay(wx.Dialog):
	...
	# ======================================================================
        # open the document with MuPDF when dialog gets created
        # ======================================================================
        # Applex modified
        **self.stream** = open(filename, "rb").read()
        self.doc = fitz.open('pdf', **self.stream**)  # create Document object

**Comment by JorjMcKie** [comment]
Glad you solved it.

And thank you for the suggestion. This would incur allocating new memory of the same amount again - in the case of your big PDF 5+ MB.

On the other hand, as became obvious here, only carrying around pointers also introduces differences in the way how a `fitz.Document` must be handled by the Python programmer ... let me think about it.

I have another place where your argument applies as well (Pixmap created from a samples area).

**Comment by JorjMcKie** [comment]
I am testing a fix for your issue.
It seems the problem can be solved, if I prevent Python freeing the area via its garbage collection, as long as the Python object `fitz.Document` lives (i.e. is not closed or deleted). This can be done by recording a reference to the bytes / bytearray object in the `Document` object.
Your example (the zip file) is working with it.

**Comment by Applex007** [comment]
Hope your fix in next version, I‘ll  close this question now

Wish you work happily!

**Comment by JorjMcKie** [comment]
@Applex007 
I am in the process of publishing v1.13.6 which fixes your issue as indicated above. The way this is implemented **does not need** an additional copy of the stream, so no extra memory load incurred.
Wheels for Linux and Mac OSX are stil being generated, but Windows wheels are already available in this repository's releases directory. You may go ahead and download the appropriate one now.

Thanks for reporting this!
